funnel/views/proposal.py

```diff
@@ -163,33 +163,25 @@ def proposal_edit(profile, space, proposal):
             _('This form uses <a href="http://daringfireball.net/projects/markdown/">Markdown</a> for formatting.')))
 
 
-# @app.route('/<space>/<proposal>/<transition>', methods=['GET', 'POST'], subdomain='<profile>')
-# @lastuser.requires_login
-# @load_models(
-#     (Profile, {'name': 'profile'}, 'g.profile'),
-#     ((ProposalSpace, ProposalSpaceRedirect), {'name': 'space', 'profile': 'profile'}, 'space'),
-#     ((Proposal, ProposalRedirect), {'url_name': 'proposal', 'proposal_space': 'space'}, 'proposal'),
-#     kwargs=True)
-# def proposal_transition(profile, space, proposal, kwargs):
-#     transition = kwargs['transition']
-#     workflow=proposal.workflow()
-
-
-# TODO: Replace this view with a state transition
-@app.route('/<space>/<proposal>/status', methods=['POST'], subdomain='<profile>')
+@app.route('/<space>/<proposal>/transition', methods=['POST',], subdomain='<profile>')
 @lastuser.requires_login
 @load_models(
     (Profile, {'name': 'profile'}, 'g.profile'),
     ((ProposalSpace, ProposalSpaceRedirect), {'name': 'space', 'profile': 'profile'}, 'space'),
     ((Proposal, ProposalRedirect), {'url_name': 'proposal', 'proposal_space': 'space'}, 'proposal'),
     permission='confirm-proposal')
-def proposal_status(profile, space, proposal):
-    form = ProposalStatusForm()
-    if form.validate_on_submit():
-        proposal._status = form.status.data
+def proposal_transition(profile, space, proposal):
+    transitionform = ProposalTransitionForm()
+    # fill up the form choices with the given proposal's transitions for validation
+    transitionform.populate_transitions(proposal)
+
+    if transitionform.validate_on_submit():  # check if the provided transition is valid
+        getattr(proposal, transitionform.transition.data)()  # call the transition
         db.session.commit()
-        flash(_("The proposal's status has changed to ") +
-              proposal.state.label.title, 'success')
+        flash(_("The proposal's status has changed to ") + proposal.state.label.title, 'success')
```

Comment:
We will now describe transitions rather than states. For all transitions, add a parameter `message='…'`. Over here, we can flash `getattr(proposal.access_for(current_auth.user), transitionform.transition.data).message`

Since we're accessing the transition object twice, might as well do `transition = getattr(…)` right after form validation.